sample/transforms.go

```diff
@@ -0,0 +1,137 @@
+package sample
+
+import (
+	"cmp"
+	"errors"
+	"math"
+	"slices"
+
+	pq "github.com/emirpasic/gods/v2/queues/priorityqueue"
+	"gonum.org/v1/gonum/floats"
+)
+
+type Transform interface {
+	Apply([]float64) ([]float64, error)
+}
+
+// TODO(parthsareen): potentially cache softmax values
+func softmax(logits []float64) []float64 {
+	var sum float64
+	probs := make([]float64, len(logits))
+	for i, v := range logits {
+		probs[i] = math.Exp(v)
+		sum += probs[i]
+	}
+	floats.Scale(1/sum, probs)
+	return probs
+}
+
+type Temperature float64
+
+func (t Temperature) Apply(logits []float64) ([]float64, error) {
```

Comment:
Is the `error` return value here required? From a brief look it seems like we could go without it, and instead provide checks elsewhere in the `sample` package before getting to the stage where `Apply` is called, keeping the function signature more straightforward

Reply:
Updating with interface changes